File: src/NTRIPServer.cpp
#include "NTRIPServer.h"

#include <WiFi.h>

#include "HandyLog.h"
#include <GpsParser.h>
#include <MyFiles.h>

#define SOCKET_RETRY_INTERVAL 30000

extern MyFiles _myFiles;

NTRIPServer::NTRIPServer(MyDisplay &display, int index)
	: _display(display), _index(index)
{
	_sendMicroSeconds.reserve(AVERAGE_SEND_TIMERS);	
}

//////////////////////////////////////////////////////////////////////////////
// Load the configurations if they exist
void NTRIPServer::LoadSettings()
{
	std::string fileName = StringPrintf("/Caster%d.txt", _index);
	// Read the server settings from the config file
	// Load last lat long
	std::string llText;
	if (_myFiles.ReadFile(fileName.c_str(), llText))
	{
		LogX(StringPrintf(" - Read config '%s'", llText.c_str()));
		auto parts = Split(llText, "\n");
		if (parts.size() > 3)
		{
			_sAddress = parts[0];
			_port = atoi(parts[1].c_str());
			_sCredential = parts[2];
			_sPassword = parts[3];
			LogX(StringPrintf(" - Recovered\r\n\t Address  : %s\r\n\t Port     : %d\r\n\t Mid/Cred : %s\r\n\t Pass     : %s", _sAddress.c_str(), _port, _sCredential.c_str(), _sPassword.c_str()));
		}
		else
		{
			LogX(StringPrintf(" - E341 - Cannot read saved Server settings %s", llText.c_str()));
		}
	}
	else
	{
		LogX(StringPrintf(" - E342 - Cannot read saved Server setting %s", fileName.c_str()));
	}
}

//////////////////////////////////////////////////////////////////////////////
// Save the setting to the file
void NTRIPServer::Save(const char *address, const char *port, const char *credential, const char *password) const
{
	std::string llText = StringPrintf("%s\n%s\n%s\n%s", address, port, credential, password);
	std::string fileName = StringPrintf("/Caster%d.txt", _index);
	_myFiles.WriteFile(fileName.c_str(), llText.c_str());
}

///////////////////////////////////////////////////////////////////////////////
// Loop called when we have new data to send
void NTRIPServer::Loop(const byte *pBytes, int length)
{
	// Disable the port if not used
	if (_port < 1 || _sAddress.length() < 1)
	{
		_status = "Disabled";
		_display.RefreshRtk(_index);
		return;
	}

	// Check the index is valid
	if (_index > RTK_SERVERS)
	{
		LogX(StringPrintf("E501 - RTK Server index %d too high", index));
		return;
	}

	// Wifi check interval
	if (_client.connected())
	{
		ConnectedProcessing(pBytes, length);
	}
	else
	{
		_wasConnected = false;
		_status = "Disconn...";
		_display.RefreshRtk(_index);
		Reconnect();
	}
}

void NTRIPServer::ConnectedProcessing(const byte *pBytes, int length)
{
	if (!_wasConnected)
	{
		_reconnects++;
		_status = "Connected";
		_display.RefreshRtk(_index);
		_wasConnected = true;
	}

	// Send what we have received
	ConnectedProcessingSend(pBytes, length);

	// Check for new data (Not expecting much)
	ConnectedProcessingReceive();
}

//////////////////////////////////////////////////////////////////////////////
// Send the data to the RTK Caster
void NTRIPServer::ConnectedProcessingSend(const byte *pBytes, int length)
{
	if (length < 1)
		return;

	// Clear out extra send history
	while (_sendMicroSeconds.size() >= AVERAGE_SEND_TIMERS)
		_sendMicroSeconds.erase(_sendMicroSeconds.begin());

	// Send and record time
	unsigned long startT = micros();
	int sent = _client.write(pBytes, length);

	unsigned long time = micros() - startT;
	if (_maxSendTime == 0)
		_maxSendTime = time;
	else
		_maxSendTime = max(_maxSendTime, time);

	if (sent != length)
	{
		LogX(StringPrintf("E500 - %s Only sent %d of %d (%sms)", _sAddress.c_str(), sent, length, time/1000));
		_client.stop();
	}
	else
	{
		// Logf("RTK %s Sent %d OK", _sAddress.c_str(), sent);
		_sendMicroSeconds.push_back(sent * 8 * 1000 / max(1UL, time));
		_wifiConnectTime = millis();
		_packetsSent++;
		_display.RefreshRtk(_index);
	}
}

//////////////////////////////////////////////////////////////////////////////
// This is usually welcome messages and errors
void NTRIPServer::ConnectedProcessingReceive()
{
	// Read the data
	int buffSize = _client.available();
	if (buffSize < 1)
		return;

	buffSize = min(buffSize, SOCKET_IN_BUFFER_MAX);

	std::unique_ptr<byte[]> buffer(new byte[buffSize + 1]);
	auto _pSocketBuffer = buffer.get();
	_client.read(_pSocketBuffer, buffSize);

	_pSocketBuffer[buffSize] = 0;

	// Log the data
	LogX("RECV. " + _sAddress + "\r\n" + HexAsciDump(_pSocketBuffer, buffSize));
}

///////////////////////////////////////////////////////////////////////////////
// Get the average send time
int NTRIPServer::AverageSendTime()
{
	if (_sendMicroSeconds.size() < 1)
		return 0;
	int total = 0;
	for (int n : _sendMicroSeconds)
		total += n;
	return total / _sendMicroSeconds.size();
}

///////////////////////////////////////////////////////////////////////////////
// Write to the debug log and keep the last few messages for display
void NTRIPServer::LogX(std::string text)
{
	auto s = Logln(text.c_str());
	_logHistory.push_back(s);
	TruncateLog(_logHistory);
	_display.RefreshRtkLog();
}

////////////////////////////////////////////////////////////////////////////////
bool NTRIPServer::Reconnect()
{
	// Limit how soon the connection is retried
	if ((millis() - _wifiConnectTime) < SOCKET_RETRY_INTERVAL)
		return false;

	_wifiConnectTime = millis();

	// Start the connection process
	LogX(StringPrintf("RTK Connecting to %s : %d", _sAddress.c_str(), _port));
	_client.setNoDelay(false);

	int status = _client.connect(_sAddress.c_str(), _port);
	_client.setNoDelay(true);					// This results in 0.5s latency when RTK2GO.com is skipped?
	if (!_client.connected())
	{
		LogX(StringPrintf("E500 - RTK %s Not connected %d. (%dms)", _sAddress.c_str(), status, millis() - _wifiConnectTime));
		return false;
	}
	LogX(StringPrintf("Connected %s OK. (%dms)", _sAddress.c_str(), millis() - _wifiConnectTime));

	if (!WriteText(StringPrintf("SOURCE %s %s\r\n", _sPassword.c_str(), _sCredential.c_str()).c_str()))
		return false;
	if (!WriteText("Source-Agent: NTRIP UM98/ESP32_T_Display_SX\r\n"))
		return false;
	if (!WriteText("STR: \r\n"))
		return false;
	if (!WriteText("\r\n"))
		return false;
	return true;
}

bool NTRIPServer::WriteText(const char *str)
{
	if (str == NULL)
		return true;

	std::string message = StringPrintf("    -> '%s'", str);
	ReplaceCrLfEncode(message);
	LogX(message);

	size_t len = strlen(str);
	size_t written = _client.write((const uint8_t *)str, len);
	if (len == written)
		return true;

	// Failed to write
	LogX(StringPrintf("Write failed %s", _sAddress.c_str()));
	return false;
}

File: WinLC29H_Server/NTRIPServer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Text;

namespace WinLC29H_Server
{
	public class NTRIPServer
	{
		private const int SOCKET_RETRY_INTERVAL_SECONDS = 300;
		private const int AVERAGE_SEND_TIMERS = 10;
		private const int SOCKET_IN_BUFFER_MAX = 1024;

		private int _index;
		private List<int> _sendMicroSeconds = new List<int>();
		private string _sAddress;
		private int _port;
		private string _sCredential;
		private string _sPassword;
		private TcpClient _client = new TcpClient();
		private NetworkStream _stream;
		private bool _wasConnected = false;
		private string _status;
		private DateTime _wifiConnectTime;
		private int _reconnects = 0;
		private int _packetsSent = 0;
		private long _maxSendTime = 0;

		/// <summary>
		/// Queue control
		/// </summary>
		Object _queueLock = new Object();
		List<byte[]> _outboundQueue = new List<byte[]>();

		public NTRIPServer(int index)
		{
			_index = index;
		}

		public bool LoadSettings()
		{
			string fileName = $"Caster{_index}.txt";
			if (!File.Exists(fileName))
				return false;
			Log.Ln("Processing " + fileName);
			var llText = File.ReadAllText(fileName);

			var parts = llText.Split('\n');
			if (parts.Length > 3)
			{
				_sAddress = parts[0].Trim('\r');
				_port = int.Parse(parts[1].Trim('\r'));
				_sCredential = parts[2].Trim('\r');
				_sPassword = parts[3].Trim('\r');
				Log.Ln($" - Recovered\r\n\t Address  : {_sAddress}\r\n\t Port     : {_port}\r\n\t Mid/Cred : {_sCredential}\r\n\t Pass     : {_sPassword}");
			}
			else
			{
				Log.Ln($" - E341 - Cannot read saved Server settings {llText}");
			}

			// Start the main processing thread
			new System.Threading.Thread(() => Loop()).Start();
			return true;
		}

		/// <summary>
		/// Process the socket connection in a seperate thread
		/// </summary>
		void Loop()
		{
			while (true)
			{
				if (_outboundQueue.Count > 0)
				{
					byte[] byteArray;
					lock (_queueLock)
					{
						byteArray = _outboundQueue[0];
						_outboundQueue.RemoveAt(0);
					}
					Loop(byteArray, byteArray.Length);
				}
				else
				{
					System.Threading.Thread.Sleep(1);
				}
			}
		}

		public void Loop(byte[] pBytes, int length)
		{
			if (_port < 1 || string.IsNullOrEmpty(_sAddress))
				return;

			if (_client.Connected)
			{
				ConnectedProcessing(pBytes, length);
			}
			else
			{
				_wasConnected = false;
				_status = "Disconn...";
				Reconnect();
			}
		}

		/// <summary>
		/// Processing of the connected socket involves sending the new data
		/// </summary>
		private void ConnectedProcessing(byte[] pBytes, int length)
		{
			if (!_wasConnected)
			{
				_reconnects++;
				_status = "Connected";
				_wasConnected = true;
			}

			ConnectedProcessingSend(pBytes, length);
			ConnectedProcessingReceive();
		}

		/// <summary>
		/// Send the data to the connected socket
		/// </summary>
		private void ConnectedProcessingSend(byte[] pBytes, int length)
		{
			if (length < 1)
				return;

			while (_sendMicroSeconds.Count >= AVERAGE_SEND_TIMERS)
				_sendMicroSeconds.RemoveAt(0);

			long startT = DateTime.Now.Ticks;
			_stream.Write(pBytes, 0, length);
			long time = DateTime.Now.Ticks - startT;

			if (_maxSendTime == 0)
				_maxSendTime = time;
			else
				_maxSendTime = Math.Max(_maxSendTime, time);

			_sendMicroSeconds.Add((int)(length * 8 * 1000 / Math.Max(1L, time)));
			_wifiConnectTime = DateTime.Now;
			_packetsSent++;
		}

		private void ConnectedProcessingReceive()
		{
			if (_stream.DataAvailable)
			{
				byte[] buffer = new byte[SOCKET_IN_BUFFER_MAX];
				int bytesRead = _stream.Read(buffer, 0, buffer.Length);
				Log.Ln("RECV. " + _sAddress + "\r\n" + HexAsciDump(buffer, bytesRead));
			}
		}

		public int AverageSendTime()
		{
			if (_sendMicroSeconds.Count < 1)
				return 0;
			int total = 0;
			foreach (int n in _sendMicroSeconds)
				total += n;
			return total / _sendMicroSeconds.Count;
		}


		/// <summary>
		/// Reconnect. But only if we have not tried for a while
		/// </summary>
		/// <returns>False if retry was not attempted</returns>
		private bool Reconnect()
		{
			if ((DateTime.Now - _wifiConnectTime).TotalSeconds < SOCKET_RETRY_INTERVAL_SECONDS)
				return false;

			_wifiConnectTime = DateTime.Now;

			Log.Ln($"RTK Connecting to {_sAddress} : {_port}");
			_client.NoDelay = false;

			try
			{
				_client.Connect(_sAddress, _port);
				_stream = _client.GetStream();
				_client.NoDelay = true;
			}
			catch (Exception ex)
			{
				Log.Ln($"E500 - RTK {_sAddress} Not connected. ({ex.Message})");
				return false;
			}

			Log.Ln($"Connected {_sAddress} OK.");

			if (!WriteText($"SOURCE {_sPassword} {_sCredential}\r\n"))
				return false;
			if (!WriteText("Source-Agent: NTRIP UM98/ESP32_T_Display_SX\r\n"))
				return false;
			if (!WriteText("STR: \r\n"))
				return false;
			if (!WriteText("\r\n"))
				return false;
			return true;
		}

		private bool WriteText(string str)
		{
			if (str is null)
				return true;

			string message = $"    -> '{str}'";
			ReplaceCrLfEncode(ref message);
			Log.Ln(message);

			byte[] data = Encoding.ASCII.GetBytes(str);
			_stream.Write(data, 0, data.Length);
			return true;
		}

		/// <summary>
		/// Dump the data as a line of hex
		/// </summary>
		private string HexAsciDump(byte[] data, int length)
		{
			var sb = new StringBuilder();
			for (int i = 0; i < length; i++)
				sb.AppendFormat("{0:X2} ", data[i]);
			return sb.ToString();
		}

		void ReplaceCrLfEncode(ref string str)
		{
			string crlf = "\r\n";
			string newline = "\\r\\n";
			str = str.Replace(crlf, newline);
		}

		/// <summary>
		/// Add this to the outbouind que for sending
		/// </summary>
		internal void Send(byte[] byteArray)
		{
			lock (_queueLock)
			{
				_outboundQueue.Add(byteArray);
			}
		}
	}
}